presto-hive/src/main/java/com/facebook/presto/hive/util/MergingPageIterator.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.util;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.SortOrder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+
+import static com.facebook.presto.hive.util.TempFileWriter.appendPositionTo;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.Iterators.concat;
+import static com.google.common.collect.Iterators.mergeSorted;
+import static com.google.common.collect.Iterators.transform;
+import static java.util.Comparator.naturalOrder;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public class MergingPageIterator
+        extends AbstractIterator<Page>
+{
+    private final List<Type> types;
+    private final List<Integer> sortFields;
+    private final List<SortOrder> sortOrders;
+    private final PageBuilder pageBuilder;
+    private final Iterator<PagePosition> pagePositions;
+
+    public MergingPageIterator(
+            Collection<Iterator<Page>> iterators,
```

Comment:
accept `PageWithPositionComparator`? This way the class will be more generic and allow for more optimized comparators.

Reply:
That interface is not accessible to connectors, and doesn't yet have any optimized implementations. It's a trivial refactoring change to do later if needed.